### ISSUE TITLE
builds.golem.network redirects to PyPI over plain HTTP for packages that do not exist, resulting in HTTP 403

### ISSUE THREAD

**Issue by cameel** [golemfactory/clay#3786]
For packages that do not exist in Golem's registry `https://builds.golem.network/simple/<package name>/` is a HTTP 302 redirect to `http://pypi.python.org/simple/<package name>/`. PyPI has disabled HTTP access quite a while ago and works only over HTTPS now. This results in a HTTP 403 error.

For that reason the latest `pip` fails to install packages from Golem's `requirements.txt`.

I think the redirect should be removed and replaced with a HTTP 404 response. Pip by default checks PyPI anyway - even when `--extra-index-url` is used. The redirect only slows it down because if it worked, `pip` would be repeating each request twice when checking for available versions.

#### Why did it work before?
This was not a problem until recently because `pip` up to version 18.0 just ignored the error as can be seen when running with `--verbose` flag:

``` bash
pip install pip==18.0
pip install --no-cache-dir --verbose --extra-index-url https://builds.golem.network appdirs
```
```
(...)
Getting page https://builds.golem.network/pip/
Starting new HTTPS connection (1): builds.golem.network:443
https://builds.golem.network:443 "GET /pip/ HTTP/1.1" 302 0
https://builds.golem.network:443 "GET /simple/pip/ HTTP/1.1" 302 0
Starting new HTTP connection (1): pypi.python.org:80
http://pypi.python.org:80 "GET /simple/pip/ HTTP/1.1" 403 16
Could not fetch URL https://builds.golem.network/pip/: 403 Client Error: SSL is required for url: http://pypi.python.org/simple/pip/ - skipping
(...)
```

`pip` 18.1 no longer ignores errors (which is good) but fails because of the HTTP redirect.

This was was originally discovered and reported in Concent. See https://github.com/golemfactory/concent-deployment/issues/296#issuecomment-455611633 for more details.

#### The redirect
``` bash
wget https://builds.golem.network/simple/appdirs/
```
```
--2019-01-18 17:30:25--  https://builds.golem.network/simple/appdirs/
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
Resolving builds.golem.network (builds.golem.network)... 188.165.220.221
Connecting to builds.golem.network (builds.golem.network)|188.165.220.221|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: http://pypi.python.org/simple/appdirs/ [following]
--2019-01-18 17:30:26--  http://pypi.python.org/simple/appdirs/
Resolving pypi.python.org (pypi.python.org)... 151.101.112.223, 2a04:4e42:1b::223
Connecting to pypi.python.org (pypi.python.org)|151.101.112.223|:80... connected.
HTTP request sent, awaiting response... 403 SSL is required
2019-01-18 17:30:26 ERROR 403: SSL is required.
```

**Comment by ktomala** [comment]
Fixed with --disable-fallback flag on pypi-server

**Comment by bartoszbetka** [comment]
It's looks like not work.
> I think the redirect should be removed and replaced with a HTTP 404 response. Pip by default checks PyPI anyway - even when --extra-index-url is used. The redirect only slows it down because if it worked, pip would be repeating each request twice when checking for available versions.

This flag disable default index:
```bash
  --disable-fallback
    disable redirect to real PyPI index for packages not found in the
    local index
```

The fix only change error from 403 to 404. I think that default index should be also support. There are places when the packages are installed from both sources(https://pypi.python.org/,https://builds.golem.network/ ) like for example: [requirements]( https://github.com/golemfactory/golem/blob/develop/apps/rendering/resources/images/requirements.txt).

**Comment by bartoszbetka** [comment]
To reproduce in newest pip `18.1`:
```bash
pip install --no-cache-dir --verbose --extra-index-url https://builds.golem.network appdirs
```
output:
```bash
Created temporary directory: /tmp/pip-ephem-wheel-cache-utr8g2g2
Created temporary directory: /tmp/pip-req-tracker-23vqvrn1
Created requirements tracker '/tmp/pip-req-tracker-23vqvrn1'
Created temporary directory: /tmp/pip-install-_s5rf7d7
Looking in indexes: https://pypi.org/simple, https://builds.golem.network
Collecting appdirs
  2 location(s) to search for versions of appdirs:
  * https://pypi.org/simple/appdirs/
  * https://builds.golem.network/appdirs/
  Getting page https://pypi.org/simple/appdirs/
  Starting new HTTPS connection (1): pypi.org:443
  https://pypi.org:443 "GET /simple/appdirs/ HTTP/1.1" 200 1212
  Analyzing links from page https://pypi.org/simple/appdirs/
    Found link https://files.pythonhosted.org/packages/c2/42/b4aaa41d488af6b265406ee90f15a30fe4f6fb5d53e0efda58e1f60e924b/appdirs-1.1.0.zip#sha256=93d276d378ecabe567a13a3a0d3a23b155dbb8a762a95acab75b2c111acb420b (from https://pypi.org/simple/appdirs/), version: 1.1.0
    Found link https://files.pythonhosted.org/packages/20/09/566235d2020da488d0628d589df1492277fe3b4eb2aa775c00a4372bc30b/appdirs-1.2.0.zip#sha256=266036a5f1eb785a5eff2a362f552b7943dc7d83737a6a201f21276bd9ef234d (from https://pypi.org/simple/appdirs/), version: 1.2.0
    Found link https://files.pythonhosted.org/packages/b3/e5/f47dce7a8357c5ef995b8d580d1e2bd0ffa658810bb307989ec209587a3b/appdirs-1.3.0.tar.gz#sha256=6a5993513c9fdf0a64260cb17b032cb563064d88d9683bab091b4672c7f3cb97 (from https://pypi.org/simple/appdirs/), version: 1.3.0
    Found link https://files.pythonhosted.org/packages/7b/8b/eebc6e2002a1e0383f1c7108d0111d4d33ea93bf417d7e19e43ec9b87b2b/appdirs-1.4.0-py2.py3-none-any.whl#sha256=85e58578db8f29538f3109c11250c2a5514a2fcdc9890d9b2fe777eb55517736 (from https://pypi.org/simple/appdirs/), version: 1.4.0
    Found link https://files.pythonhosted.org/packages/bd/66/0a7f48a0f3fb1d3a4072bceb5bbd78b1a6de4d801fb7135578e7c7b1f563/appdirs-1.4.0.tar.gz#sha256=8fc245efb4387a4e3e0ac8ebcc704582df7d72ff6a42a53f5600bbb18fdaadc5 (from https://pypi.org/simple/appdirs/), version: 1.4.0
    Found link https://files.pythonhosted.org/packages/78/14/0b89ddf22e6466d4bb46c718b3f2fe3b2fb1060d0ce1094a06e4aa473a39/appdirs-1.4.1-py2.py3-none-any.whl#sha256=566f148ff9f2e9a36912637a4ab1eaf708a499255d42a055c59dd0549507583d (from https://pypi.org/simple/appdirs/), version: 1.4.1
    Found link https://files.pythonhosted.org/packages/7c/26/b6b1222f79a56bc96a705eae343ec48fb3725c530c31af2133a3c4a33b2c/appdirs-1.4.1.tar.gz#sha256=95259bccef631e5e44438c50087f761d699005e472f8f77f20ec968e51a7e10e (from https://pypi.org/simple/appdirs/), version: 1.4.1
    Found link https://files.pythonhosted.org/packages/ff/f8/7be9a3ff64bb18cb8f8fdcbd4bb4097903f2e4f930a337253f85325076c5/appdirs-1.4.2-py2.py3-none-any.whl#sha256=a53330b9d53b66aba1e26907dea2958982ebad92735f9faf3897b73c909a20c1 (from https://pypi.org/simple/appdirs/), version: 1.4.2
    Found link https://files.pythonhosted.org/packages/88/99/293dac0b3cdf58ce029ec5393624fac5c6bde52f737f9775bd9ef608ec98/appdirs-1.4.2.tar.gz#sha256=e2de7ae2b3be52542b711eacf4221683f1d2f7706a5550cb2c562ee4ba93ee74 (from https://pypi.org/simple/appdirs/), version: 1.4.2
    Found link https://files.pythonhosted.org/packages/56/eb/810e700ed1349edde4cbdc1b2a21e28cdf115f9faf263f6bbf8447c1abf3/appdirs-1.4.3-py2.py3-none-any.whl#sha256=d8b24664561d0d34ddfaec54636d502d7cea6e29c3eaf68f3df6180863e2166e (from https://pypi.org/simple/appdirs/), version: 1.4.3
    Found link https://files.pythonhosted.org/packages/48/69/d87c60746b393309ca30761f8e2b49473d43450b150cb08f3c6df5c11be5/appdirs-1.4.3.tar.gz#sha256=9e5896d1372858f8dd3344faf4e5014d21849c756c8d5701f78f8a103b372d92 (from https://pypi.org/simple/appdirs/), version: 1.4.3
  Getting page https://builds.golem.network/appdirs/
  Starting new HTTPS connection (1): builds.golem.network:443
  https://builds.golem.network:443 "GET /appdirs/ HTTP/1.1" 302 0
  https://builds.golem.network:443 "GET /simple/appdirs/ HTTP/1.1" 404 None
Could not install packages due to an EnvironmentError.
Traceback (most recent call last):                                                                                                                                                                                   
  File "/usr/lib/python3/dist-packages/pip/_internal/commands/install.py", line 338, in run                                                                                                                          
    resolver.resolve(requirement_set)                                                                                                                                                                                
  File "/usr/lib/python3/dist-packages/pip/_internal/resolve.py", line 102, in resolve                                                                                                                               
    self._resolve_one(requirement_set, req)                                                                                                                                                                          
  File "/usr/lib/python3/dist-packages/pip/_internal/resolve.py", line 256, in _resolve_one                                                                                                                          
    abstract_dist = self._get_abstract_dist_for(req_to_install)                                                                                                                                                      
  File "/usr/lib/python3/dist-packages/pip/_internal/resolve.py", line 209, in _get_abstract_dist_for                                                                                                                
    self.require_hashes                                                                                                                                                                                              
  File "/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py", line 218, in prepare_linked_requirement                                                                                                 
    req.populate_link(finder, upgrade_allowed, require_hashes)                                                                                                                                                       
  File "/usr/lib/python3/dist-packages/pip/_internal/req/req_install.py", line 164, in populate_link                                                                                                                 
    self.link = finder.find_requirement(self, upgrade)                                                                                                                                                               
  File "/usr/lib/python3/dist-packages/pip/_internal/index.py", line 572, in find_requirement                                                                                                                        
    all_candidates = self.find_all_candidates(req.name)                                                                                                                                                              
  File "/usr/lib/python3/dist-packages/pip/_internal/index.py", line 530, in find_all_candidates                                                                                                                     
    for page in self._get_pages(url_locations, project_name):                                                                                                                                                        
  File "/usr/lib/python3/dist-packages/pip/_internal/index.py", line 675, in _get_pages                                                                                                                              
    page = self._get_page(location)                                                                                                                                                                                  
  File "/usr/lib/python3/dist-packages/pip/_internal/index.py", line 793, in _get_page                                                                                                                               
    return _get_html_page(link, session=self.session)                                                                                                                                                                
  File "/usr/lib/python3/dist-packages/pip/_internal/index.py", line 147, in _get_html_page                                                                                                                          
    resp.raise_for_status()                                                                                                                                                                                          
  File "/usr/share/python-wheels/requests-2.20.0-py2.py3-none-any.whl/requests/models.py", line 940, in raise_for_status                                                                                             
    raise HTTPError(http_error_msg, response=self)                                                                                                                                                                   
pip._vendor.requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://builds.golem.network:443/simple/appdirs/                                                                                     
Cleaning up...
Removed build tracker '/tmp/pip-req-tracker-23vqvrn1'

```

**Comment by mfranciszkiewicz** [comment]
This has been fixed in pypiserver 1.2.2:
https://github.com/pypiserver/pypiserver/releases/tag/v1.2.2

`builds.golem.network` is running v1.2.0

**Comment by cameel** [comment]
Do you mean this?

> FIX: update fallback URL to https://pypi.org/simple since pypi.python.org
has shut down

If I understand correctly, this will make the server redirect to the HTTPS URL rather than the HTTP one. So this will work around the HTTP 404 problem reported by @bartoszbetka above if @ktomala updates to 1.2.2 and re-enables the redirect.

The thing is that the HTTP 404 should not be a problem. It's normal for a package not to always be present in all the registries. I've found an old, but still open issue in `pip`'s bug tracker: https://github.com/pypa/pip/issues/4580. It describes the exact same problem, though between older versions (8.1.2 and 9.0.1). I think it's a bug in `pip` so I dropped a comment there.

The people in the issue say that it happens only when using `--no-cache-dir`. And indeed for us this workaround works too. But it's not ideal because when building Docker images we don't want the cache. The workaround for the workaround is to remove the cache manually after installing packages.

So, to summarize:
- @ktomala has disabled the HTTP redirect which is the right solution in my opinion but sadly `pip` seems to have another bug that makes it unusable without the redirect.
- Not using `--no-cache-dir` is one of the workarounds.
- Keeping the redirects on the server is another workaround that should work, but requires upgrading to `pypiserver` 1.2.2.

If we want to keep the redirects then perhaps we should replace `--extra-index-url` with `--index-url` in requirements? Because with redirects Golem's repository can serve as a proxy for the main repository and having pip check the main repository in addition to that is redundant. It would only be useful if we wanted to have packages that have some versions available in Golem's repository and others on PyPI.

**Comment by ktomala** [comment]
Updated to 1.2.5
